crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -0,0 +1,188 @@
+            self.metrics
+                .precompile_cache_size
+                .increment((new_entry_count - previous_entry_count) as f64);
+        }
+    }
+}
+
+impl Precompile for CachedPrecompile {
+    fn call(&self, data: &Bytes, gas_limit: u64) -> PrecompileResult {
+        let key = CacheKey(data.clone());
+
+        let cache_result = self.cache.get(&key);
+
+        if let Some(ref entry) = cache_result {
+            match &entry.result {
+                Ok(_) => {
+                    self.increment_by_one_precompile_cache_hits();
+                    if gas_limit < entry.gas_used {
+                        return Err(PrecompileError::OutOfGas)
+                    }
+                    return entry.result.clone()
+                }
+                Err(PrecompileError::OutOfGas) => {
+                    if gas_limit <= entry.err_gas_limit {
+                        // entry.err_gas_limit gave OOG previously, anything equal or below
+                        // should fail with OOG too.
+                        self.increment_by_one_precompile_cache_hits();
+                        return Err(PrecompileError::OutOfGas)
+                    }
+                }
+                err => {
+                    if gas_limit > entry.err_gas_limit {
+                        // entry.err_gas_limit did not give OOG previously (it gave `err`),
+                        // anything equal or above should not give it either
+                        self.increment_by_one_precompile_cache_hits();
+                        return err.clone()
+                    }
+                }
+            }
+        }
+
+        // cache miss or error with not yet seen gas limit, call the precompile
+        self.increment_by_one_precompile_cache_misses();
+        let result = self.precompile.call(data, gas_limit);
+
+        let previous_entry_count = self.cache_entry_count();
+        match &result {
+            Ok(val) => {
+                self.cache.insert(
+                    key,
+                    CacheEntry { result: result.clone(), gas_used: val.gas_used, err_gas_limit: 0 },
+                );
+            }
+            _ => {
+                self.cache.insert(
+                    key,
+                    CacheEntry { result: result.clone(), gas_used: 0, err_gas_limit: gas_limit },
+                );
+            }
+        }
+
+        self.update_precompile_cache_size(previous_entry_count);
+        result
+    }
+}
+
+/// Metrics for the cached precompile provider, showing hits / misses for each cache
+#[derive(reth_metrics::Metrics, Clone)]
+#[metrics(scope = "sync.caching")]
+pub(crate) struct CachedPrecompileMetrics {
+    /// Precompile cache hits
+    precompile_cache_hits: metrics::Gauge,
+
+    /// Precompile cache misses
+    precompile_cache_misses: metrics::Gauge,
+
+    /// Precompile cache size
+    ///
+    /// NOTE: this uses the moka caches' `entry_count`, NOT the `weighted_size` method to calculate
+    /// size.
```

Comment:
we should do this right? since inputs can have different sizes? Or are the limits fairly conservative right now